aggregator/src/test/java/org/jenkinsci/plugins/workflow/BuildQueueTasksTest.java

```diff
@@ -0,0 +1,192 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow;
+
+import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowExecution;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.test.steps.SemaphoreStep;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runners.model.Statement;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.RestartableJenkinsRule;
+import org.xml.sax.SAXException;
+
+import static org.junit.Assert.*;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+
+import com.gargoylesoftware.htmlunit.Page;
+
+import hudson.model.Result;
+import hudson.model.queue.QueueTaskFuture;
+import net.sf.json.JSONArray;
+import net.sf.json.JSONObject;
+
+public class BuildQueueTasksTest {
+
+    @Rule public RestartableJenkinsRule story = JenkinsRuleExt.workAroundJenkins30395Restartable();
+
+    @Issue("JENKINS-28649")
+    @Test public void queueAPI() throws Exception {
+        // This is implicitly testing ExecutorStepExecution$PlaceholderTask as exported bean
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                // use non-existent node label to keep the build queued
+                p.setDefinition(new CpsFlowDefinition("node('linux') { echo 'test' }"));
+
+                WorkflowRun b = scheduleAndWaitQueued(p);
+                assertQueueAPIStatusOKAndAbort(b);
+            }
+        });
+    }
+
+    @Issue("JENKINS-28649")
+    @Test public void queueAPIRestartable() throws Exception {
+        // This is implicitly testing AfterRestartTask as exported bean
```

Comment:
I can not imagine how to test `AfterRestartTask` (unless mocking a lot of things to insert wait points). If you have some straightforward way to do it in mind, please tell me, I'll be happy to implement it. If not, sorry, I give up.